### PR TITLE
Add Node concept and start network structure page.

### LINKED ISSUE #19: The "Node" concept
I think it could be useful for describing the details of how we approach distributed operations to introduce the concept of a "Node" to express a locally encapsulated (subset of) an ECP deployment.

This could also be useful for what was proposed in the discussion on reliability (#4). That is, to have an alternative (less complicated, more reliable) message transport mode for "local" (i.e. limited to one Node) applications, e.g. using Queues between threads instead of conventional zmq.
In the same breath, it makes sense to also introduce the term "message transport" (name TBC, I'm open for improvement suggestions)
Already refined formulation:

> Node
> A Node is a local context in which (part of) an ECP deployment runs. This may be a single application using one or more threads or processes. An ECP network has one or more Nodes. If it has a single Node, its Components may use the Local Message Transport ("local mode"). If it has multiple Nodes, they must use the Distributed Message Transport ("distributed mode").
> (TBC) Optional feature "Bridging Coordinator": In a DMT network, if the Components of a Node use local mode, only a Coordinator in that Node may use DMT to bridge messages to/from other Nodes. Put differently, Components in local mode may only communicate to outside their Node via a Coordinator in their Node.
>
> Message Transport (Distributed/Local)
> The communication layer that transports ECP messages between Components. The Local Message Transport (default TBD) only works within a Node. The Distributed Message Transport (zeromq tcp protocol) also works across Nodes. Local Message Transport options include zeromq inproc, zeromq IPC, and queues between threads/processes (TBC).

~~Open question: In some contexts, "node" means a (physical) computer. Does it make sense to preserve/integrate that interpretation, or does it not make sense as for zmq it doesn't really matter if sources/sinks are on the same PC or somewhere else over the network? (or does it?)~~ Let's scratch that.

Thoughts?

### PR DESCRIPTION
**Pull request by bilderbuchi**
Closes #19, #36, #37.

It shows that defining one concept (Nodes) drags in a couple of others that need to defined, too.
Feedback, please!

### CODE REVIEW

File: network-structure.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+:::{admonition} TODO
+Decide if the "mode" is the same for all Components in a Node, and if yes, how we track which Node a Component belongs to, and how the mode information propagates within.
+:::
+
+## Message Layer
+The Message Layer is the communication layer that concerns itself with ECP message (de)composition, validation, serialisation, etc.
+
+## Transport Layer
+The Transport Layer is the communication layer that transports ECP messages between Components.
+This uses zeromq or simpler localised methods, see the next section "Message Transport Mode".
```
Comment (bmoneke) [inline network-structure.md:23]: I'm unsure, whether we understand the same things. In my opinion:

Transport Layer is not only zeromq (or other), but also contains the routing information in one node or between nodes. So #22 is part of the transport layer.

Do you understand "message" as the transported payload (then I agree with your definitions) or as the whole communicated data (including addresses etc.)?

I'd define, for clarification:
```suggestion
## ECP Message
An ECP Message is one set of data transmitted from one Component to another one. It consists of a Header, necessary to deliver the Message to its destination, and of some Content transmitted.

## Message Layer
The Message Layer is the communication layer that concerns itself with (de)composition, validation, serialisation, etc. of ECP Message Contents.

## Transport Layer
The Transport Layer is the communication layer that transports ECP messages between Components.
This uses zeromq or simpler localised methods, see the next section "Message Transport Mode".
```

File: network-structure.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+An ECP network has one or more Nodes. 
+If it has a single Node, its Components may use the Local Message Transport (LMT/"local mode"), see below. 
+If it has multiple Nodes, they must use the Distributed Message Transport (DMT/"distributed mode").
+
+(TBC) Optional feature "Bridging Coordinator": In a DMT network, if the Components of a Node use local mode, only a Coordinator in that Node may use DMT to bridge messages to/from other Nodes. Put differently, Components in local mode may only communicate to outside their Node via a Coordinator in their Node.
+:::{admonition} TODO
+Decide if we include the Bridging Coordinator concept (I think we will).
+:::
+:::{admonition} TODO
+Decide if the "mode" is the same for all Components in a Node, and if yes, how we track which Node a Component belongs to, and how the mode information propagates within.
```
Comment (bmoneke) [inline network-structure.md:15]: I'd say, that you can have both modes in the same node: If you tie Components to the local Coordinator, they use the local mode, or they can participate in the DMT (and talk to other Components in their node via the DMT and through the local Coordinator from the "outside").

One "implementation" idea, could be, that you give a socket to a Component, which can either be a zmq socket to a DMT Coordinator or some object connected (maybe using pipes, queues, but with the same methods as a zmq socket) to the local Coordinator.

Basically you give a different "adapter" (in pymeasure lingo).

File: network-structure.md (hunk 1 above)
Comment (bilderbuchi) [reply network-structure.md:23]: > Transport Layer is not only zeromq (or other), but also contains the routing information in one node or between nodes. So https://github.com/pymeasure/experiment_control_protocol/issues/22 is part of the transport layer.

Good point!

> Do you understand "message" as the transported payload (then I agree with your definitions) or as the whole communicated data (including addresses etc.)?

The whole data, header, payload, checksum.

File: network-structure.md (hunk 2 above)
Comment (bilderbuchi) [reply network-structure.md:15]: Hm, maybe we need to "pivot" this "mode" concept a little?
What I want to express/enable is that we have a couple of routing options:
* Use the Node-local Coordinator to route messages anywhere in the network (DMT, has to use zeromq tcpip)
* Use the Node-local Coordinator to route messages anywhere in the local Node (DMT, has to use zeromq tcpip)
* If you want (e.g. for increased reliability, performance, minimized network/port use...), you may: Use the Node-local Coordinator to route messages anywhere in the local Node (LMT, use zeromq inproc, queues or one of the other local options)

I would say that a Coordinator is exceptional, as it always needs one DMT connection (the one to the other Coordinators). The question is for the Node-facing "side" of the Coordinator (which can either be DMT or LMT): Should it be allowed to mix-and-match between LMT and DMT connections to its Components? Does allowing mixed use make the spec/logic much more complicated (for not much gain)? 
If it does not restrict us too much, I'd suggest to disallow mixing -- we can always relax that later, while it will be more difficult to start with mixed use and restrict it later.

File: network-structure.md (hunk 1 above)
Comment (bilderbuchi) [reply network-structure.md]: I pushed an update.

File: network-structure.md (hunk 2 above)
Comment (bmoneke) [reply network-structure.md:15]: If you have another dmt Coordinator running, I do not see the benefit, that the node's local coordinator offers dmt connection to its components.

But it should not be too complex: The Coordinator polls all connections for incoming messages and handles them.
If there is one socket more or less, does not matter as long as the messages have the same format.

Comment (bmoneke) [reply network-structure.md:15]: Let us restrict it (hinting at the possibility to open it later), then we try to mix it experimentally. According to the results, we can allow a mix or not.

Comment (bilderbuchi) [reply network-structure.md:15]: > If you have another dmt Coordinator running, I do not see the benefit, that the node's local coordinator offers dmt connection to its components.

I made this point elsewhere already I think, but one advantage to use the Node-local Coordinator is closeness -- that "other DMT Coordinator" could be on another machine, which is slower and less reliable. 
_Should I add a note on this in the document?_
We could even specify that Components need to use their own Node's Coordinator.

Comment (bilderbuchi) [reply network-structure.md]: OK, I tightened up usage a bit (LMT/DMT choice per Node, exactly one Coordinator per Node). We can always relax later should we need it, but I want to try to keep complexity down so we don't get lost in details.

File: network-structure.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+Components within a Node may use the Local Message Transport (LMT/"local mode") or the Distributed Message Transport (DMT/"distributed mode") to communicate with each other, see below.
+All Components of a Node must use the same Message Transport Mode.
+:::{admonition} Note
+In the future this might be relaxed, so that Components within a Node might operate in different modes.
+:::
+:::{admonition} TODO
+Decide how we track which Node a Component belongs to, and how the mode information propagates within a Node.
+:::
+
+### Bridging Coordinators
```
Comment (bmoneke) [inline network-structure.md:24]: Let us move this part out of a node, maybe a general comment is better.
There we can state, that every Component (except for Coordinators) is connected to exactly one Coordinator. The Coordinators are connected to each other, such that any Component may send a message to any other Component via their respective Coordinators.

Comment (bilderbuchi) [reply network-structure.md:24]: Good idea, I modified this a bit (all your relationships are actually one-on-one, a flowchart offers the possibiltiy "Node" subgraphs.
```mermaid
flowchart LR

    subgraph Node1
        Component1 -- "connects to" --- Coordinator1
        Component2 -- "connects to" --- Coordinator1
    end
    Coordinator1 -- "coordinates with" --- Coordinator2
    subgraph Node2
        Coordinator2 -- "connects to" --- Component3
        Coordinator2 -- "connects to" --- Component4
    end
```

Comment (bmoneke) [reply network-structure.md]: Nice.

**Review by bklebel — CHANGES_REQUESTED** [review]
Thanks for this! 

Most of this PR is fine with me, however we, it seems almost by accident, would decide that we need a separate Coordinator on every machine which we use (i.e. a Node can only be working one machine by design). This is a discussion we might want to have in a separate issue - it touches fundamentally the entry threshold, and future configuration effort. 
If we keep out statements that "every Component must connect to one and only one (i.e. 'exactly one') Coordinator", and "every node must have one and only one Coordinator", the rest would be fine by me.

**Comment by bklebel** [comment]
One more comment on this: You people are amazingly crazy fast. It's a pleasure to work with you!

File: components.md [hunk 4]
```diff
@@ -94,11 +94,18 @@ sequenceDiagram
 :::
 
 ## Coordinator
-A component primarily concerned with routing/coordinating the message flow between other Components.
+A component primarily tasked with routing/coordinating the message flow between other Components.
 It represents the intermediate zmq brokers, proxies or somesuch.
```
[the remark this answers is not in the document]
Comment (bilderbuchi) [reply components.md:98]: Depending on how we decide to wire the logging and data protocols, _all_ 3 might be in the Coordinator, hence the plural. Also "be" is I think too strong a formulation - it implies an "is-a" relationship in the OOP sense, i.e. inheritance, and I want to avoid that.

File: components.md [hunk 5]
```diff
@@ -94,11 +94,18 @@ sequenceDiagram
...
+
+Every network needs at least one Coordinator, as all other Components within a Node directly exchange messages only with the Coordinator.
+
+Every Node must have exactly one Coordinator.
```
[the remark this answers is not in the document]
Comment (bilderbuchi) [reply components.md:102]: > Thus, even if we have only 1 Actor, 1 Director and 1 Observer, we might need 1 Coordinator for the control channel (heavily discussed with DEALER and ROUTER socket comms), 1 Coordinator for the data stream (rather simple zmq proxy with XPUB and XSUB sockets, to enable n-to-n publishing) 1 Coordinator for the logging stream (same thing as the data stream, but not the same entity, i.e. not the same ports at least), which are 3 Coordinators.

My assumption when writing this was that all 3 channels (control, data, logging) are coordinated by a Coordinator, which I did not see and disadvantages of. I guess we'll have to discuss that further (elsewhere).

[the remark this answers is not in the document]
Comment (bilderbuchi) [reply components.md:102]: > for every Node (i.e. every machine involved)

I think we'll have to iterate on the formulation of "Node". The way I understand and intend is _not_ to restrict one Node to one machine (hence the "may" in "This may be a single application using one or more threads or processes."). A Node should be the "local" context in the sense that 1 Coordinator and N other Components make up a local "community"/graph, the Node (irrespective of location). 
We need to expand on the understanding of "local" here.

One wrinkle that this uncovers (thank you!) is that we want to enable a Node to use thread queues in LMT, which is of course not possible with its Components distributed between machines. 
We should be able to cover this with additional restrictions on LMT usage. Maybe key the restrictions on the individual capabilities of the various methods, and explicitly enumerate the methods.

**Comment by bilderbuchi** [comment]
> One more comment on this: You people are amazingly crazy fast. It's a pleasure to work with you!

Thanks! ❤️ Yesterday afternoon was quite intense, I was just constantly working off Github notifications from incoming comments on my comments. We should really schedule a video call in our matrix room...

File: components.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (bklebel) [reply components.md:102]: > all 3 channels (control, data, logging) are coordinated by a Coordinator, which I did not see and disadvantages of. 

It seems I have missed out on some other discussions (i.e. #32), I will have to read up on them to catch how new terms are used. Just for me, the term "Coordinator" was "one entity which helps to enable the communication", and multiple entities might be needed for successful working with multiple channels. It is no problem (for me) to stuff all needed Coordinators multithreaded in a single process for simplicity, but they are still separate things. 

Also, I thought the _whole deal_ of the LMT Node concept was to have it fully local, independent of zmq and network infrastructure, e.g. with queues. As far as I understood [queues](https://docs.python.org/3/library/queue.html) in python, they are all inside one process, which would be fine for that purpose. 

> A Node should be the "local" context in the sense that 1 Coordinator and N other Components make up a local "community"/graph, the Node (irrespective of location). 

Ok, with the additional behavior that one Node is not restricted to one machine, but serves as network delimiter/namespace or similar, this is fine for me, if we give one Node "one and only one control-channel Coordinator". I don't think we would like to have multiple data-channel Coordinators (proxies) by design at all, e.g. with one per Node, because one big deal of having this one proxy in one central place would be for Observers to only listen at one point. Like this we do not need to start with issues like multi-hop-like distribution of the data, as we are discussing things for the control-Coordinators. If we would ever need an additional data-channel proxy, this would be for performance issues, I think, and then I don't really think that we would get this kind of issue within the scope of our goals.

[the remark this answers is not in the document]
Comment (bilderbuchi) [reply components.md:102]: > Also, I thought the whole deal of the LMT Node concept was to have it fully local, independent of zmq and network infrastructure, e.g. with [queues](https://docs.python.org/3/library/queue.html). As far as I understood queues in python, they are all inside one process, which would be fine for that purpose.

Initially yes, but then it turned out that zmq can also do the "local" stuff (ipc, inproc), so it seems good to add that to the LMT options. Maybe we should limit LMT to queues and zmq inproc, and if a node is in LMT, it consequently has to live in one process. What do you think?

[the remark this answers is not in the document]
Comment (bilderbuchi) [reply components.md:102]: > I don't think we would like to have multiple data-channel Coordinators (proxies) by design at all, e.g. with one per Node, because one big deal of having this one proxy in one central place would be for Observers to only listen at one point. Like this we do not need to start with issues like multi-hop-like distribution of the data, as we are discussing things for the control-Coordinators.

Darn it, of course, I hadn't thought of that, yet. I think I agree.
We could make it so our "Coordinator" is purely about the control channel, and have another (singleton) Component be about data proxying (and maybe logging at the same time, as it's similar). I'll open an issue.

File: components.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (bilderbuchi) [reply components.md:98]: I used your wording now.

File: network-structure.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+All Components of a Node must use the same Message Transport Mode.
+:::{admonition} Note
+In the future this might be relaxed, so that Components within a Node might operate in different modes.
+:::
+:::{admonition} TODO
+Decide how we track which Node a Component belongs to, and how the mode information propagates within a Node.
+:::
+
+## Multi-Node networks
+Every Component (except for Coordinators) is connected to exactly one Coordinator.
```
[the remark this answers is not in the document]
Comment (bilderbuchi) [reply network-structure.md]: I think this is clearer now.

File: network-structure.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+In the future this might be relaxed, so that Components within a Node might operate in different modes.
+:::
+:::{admonition} TODO
+Decide how we track which Node a Component belongs to, and how the mode information propagates within a Node.
+:::
+
+## Multi-Node networks
+Every Component (except for Coordinators) is connected to exactly one Coordinator. 
+The Coordinators are connected to each other, such that any Component may only send a message to any other Component via their respective Coordinators.
+Put differently, Components may only communicate to outside their Node via a Coordinator in their Node.
```
[the remark this answers is not in the document]
Comment (bilderbuchi) [reply network-structure.md]: I think this still reflects our current understanding of the situation. If not, could you specify what you would like changed? Or would you merge it as-is?

**Review by bmoneke — APPROVED** [review]
Looks good.

Thanks for the work.

@bklebel, normally the person who opened a comment/conversation should mark it as resolved, when it is resolved in his opinion.
Therefore, @bilderbuchi left the comments open.

N. B. It only works when you have write permission. It is however useful, that the raising party has the control, whether the issue is resolved or not.

**Review by bklebel — APPROVED** [review]
LGMT

**Comment by bilderbuchi** [comment]
🥳
